### PR TITLE
Better viewing

### LINKED ISSUE #377: Allow `cell.comp(0)` or `net.branch(0)` or `net.comp(0)`
Should we allow `cell.comp(0)` or `net.branch(0)` or `net.comp(0)`? I think this would be helpful sometimes to be able to reference comps by their global index that can also be found in `cell.nodes`. Currently this can be achieved via cell.branch.comp(0), which I think might be an artefact and is less clear.

### PR DESCRIPTION
**Pull request by jnsbck**
I have been working on this for a while, but I think I have figured 99% of the difficult things out, so it's time to make a PR to get your feedback on it and to slowly integrate it into `main`. I suspect it will take a while since it messes with a lot of stuff under the hood, but in the long run it should make life much easier.

The aim of this PR is to change how `View` works in `Jaxley`. At the moment `Module` and `View` are two different objects and hence they behave very differently. For example a `View` cannot be integrated. Under the hood treating `Module` and `View` differently also leads to a lot of duplicate methods and boilerplate since all methods that we want to have in both `Module` and `View` have to be implemented in `Module` and wrapped once to expose them for `Module` and once to expose them for `View` via the pointer that points back to `Module`. I therefore propose to handle viewing differently, namely:
1. Make `View` a sublcass of `Module`. This means if `Module` returns a `View` all attrs and methods still work.
2. `View` controls how the attrs of `Module` are presented, i.e. `cell.branch(0).nodes` will return a dataframe with the subset of nodes that are part of branch 0. Which part of a `Module` is in view is kept track of internally by an index. 
3. Compute global and local indexes at init and add them to `nodes`.

Since `View` is a `Module`, viewing is recursive, i.e. `View(View(Module))` works as well. Furthermore and importantly, the current API stays the same for the user.

This is very cool, because it allows for a lot of new features:

Indexing is now more flexible, we could i.e. sample them and index into the Module using `at` or index into comps straight from `Network`.
```python
# more flexible indexing / selection with at 
# (tracked via dataframe index, which does not change with scope)
rnd_inds = np.random.randint(0, len(net.nodes),10)
net.at(rnd_inds)

# arbitrary selection
net.branch(0).show()
net.comp(0).show()

# lazy indexing
net[0,0,0].show()

# shape
net.shape
net.cell(0).shape
```

Additionally, keeping track of global and local indexes in nodes allows to select which "scope" to use when indexing.
```python
# scope
net.set_scope("global")
net.cell([0,2]).branch([0]).comp([1,2]).show() # -> [1,2] 

net.set_scope("local")
net.cell([0,2]).branch([0]).comp([1,2]).show() # -> [1,2,41,42]

net.scope("local").comp(0).show()
# vs.
net.scope("global").comp(0).show()
```

Its now also easy to add context management and iterables.
```python
# context management
with net.cell(0).branch(0).comp(0) as comp0:
    comp0.set("v", -70)
    comp0.set("radius", 0.1)
net.cell(0).branch(0).comp([0,1]).show()[["v", "radius"]]

# iterables
for cell in net.cells:
    for branch in cell.branches:
        for comp in branch.comps:
            comp.set("v", -71)

for comp in net.cell(0).branch(0).comps:
    comp.set("v", -72)
net.show()[["v"]]
```

Groups, channels and synapses now basically behave the same and can be called at any time.
```python
# groups
net.cell(1).branch(0).add_group("group")
net.group.show()

# Channel and Synapse views
net.HH.show()
net.cell(0).HH.nodes
net.HH.cell(0).nodes

net.TestSynapse.nodes
net.cell([0,1]).TestSynapse.nodes
net.TestSynapse.cell(0).nodes
```

And indexing into edges/synapses is now also possible using `edge`
```python
# edges
net.edge([0,1,2]).edges
```

And arguably the coolest feature imo is being able to copy / extract elements of a Module, since Views <=> Module, which could be used to simulate only parts of a network for example.
```python
# copying
cell0 = net.cell(0).copy()
cell0.show()
```

Very curious to hear your feedback or suggestions you have. If you wanna play around with this yourselves, you can pull from this branch.

Merging this will make #329, #330, #351 obsolete and close #377, #354, probably #338 and I guess also #133
Will also make #264 and #134 trivial
Also might wanna revisit #291 after this

### CODE REVIEW

**Comment by jnsbck** [comment]
I am in the process of integrating this into the rest of Jaxley now. I am focusing on getting all tests to pass first. Surprisingly many already pass 110/175, as of writing this.

### Todos:
- [ ] Fix broken tests
- [ ] Take care of #TODOs
- [ ] Add new tests
- [ ] Add documentation and type hints
- [ ] Write documentation text
- [ ] Add remaining attrs of Module to View

I suspect with some more planning and smart use of the new functionalities, this will make the existing code much nicer and probably a bunch of things obsolete. 
- [ ] Remove obsolete code
- [ ] More careful refactor using the new functionalities to reduce line count


### Comments:
- Moved `_append_multiple_synapses` and associated functionality to `Network`.
- `SynapseView` replaced by `module.TestSynapse.edge([0,1])`

### Open Questions:
- How should we name `self.view`, `self.base`, `self.at`?

**Comment by jnsbck** [comment]
123/174

**Comment by jnsbck** [comment]
153/174

**Comment by jnsbck** [comment]
159/174

**Comment by jnsbck** [comment]
165/174

**Comment by jnsbck** [comment]
168/174

**Comment by jnsbck** [comment]
174/174 Omg all tests are passing! :partying_face: :tada:

**Comment by michaeldeistler** [comment]
Wohooo amazing!!!

**Comment by jnsbck** [comment]
All tests are passing and I have added a decent amount of documentation, so you should be able to do a first round of reviewing. There is still some open todos and questions, some of which will probably need their own PR. Lemme know if anything requires further explanation.

File: jaxley/modules/base.py [hunk 1]
```diff
@@ -1697,578 +2046,314 @@ def _rotate(self, degrees: float, rotation_axis: str, view: pd.DataFrame):
...
+            where_comps = base_nodes["global_comp_index"].isin(incl_comps)
+            possible_nodes_in_view = base_nodes.index[where_comps].to_numpy()
+            self._nodes_in_view = np.intersect1d(
+                possible_nodes_in_view, self._nodes_in_view
+            )
+        elif has_node_inds and has_edge_inds:
+            self._nodes_in_view = at_nodes
+            self._edges_in_view = at_edges
+
+    # TODO: Implement this
```
Comment (michaeldeistler) [inline jaxley/modules/base.py]: remove the code for now? Or at least add a docstring what this should do and why it would be good to have.

**Review by michaeldeistler — CHANGES_REQUESTED** [review]
This is super awesome! Comments below, also happy to chat. Let's not rename anything yet, we can really also discuss this tomorrow in person and then decide.

File: jaxley/modules/base.py [hunk 2]
```diff
@@ -120,8 +123,137 @@ def __init__(self):
...
+        I.e. for net -> [cell, branch, comp]. For branch -> [comp]"""
+        levels = ["network", "cell", "branch", "comp"]
+        view_lvl = self._current_view
+        children = levels[levels.index(view_lvl) + 1 :]
+        return children
+
+    def __getitem__(self, index):
+        supported_lvls = ["network", "cell", "branch"]  # cannot index into comp
+
+        # TODO: SHOULD WE ALLOW GROUPVIEW TO BE INDEXED?
```
[the remark this answers is not in the document]
Comment (jnsbck) [reply jaxley/modules/base.py:182]: I meant in this context, i.e. using lazy indexing.

File: jaxley/modules/base.py [hunk 3]
```diff
@@ -159,19 +294,288 @@ def _update_nodes_with_xyz(self):
...
+            scope: either "global" or "local".
+
+        Returns:
+            View with the specified scope."""
+        view = self.view
+        view.set_scope(scope)
+        return view
+
+    def _at_level(self, level: str, idx: Any) -> View:
+        """Return a View of the module at the specified level and index.
```
[the remark this answers is not in the document]
Comment (jnsbck) [reply jaxley/modules/base.py:400]: also disliked this somehow. changed to `_at_nodes(self, key, idx)` now.

File: jaxley/modules/base.py [hunk 4]
```diff
@@ -159,19 +294,288 @@ def _update_nodes_with_xyz(self):
...
+        """
+        next_level = self._viewing_levels()[0]
+        yield from self._iter_level(next_level)
+
+    @property
+    def shape(self) -> Tuple[int]:
+        """Returns the number of submodules contained in a module.
+
+        ```
+        network.shape = (num_cells, num_branches, num_compartments)
```
[the remark this answers is not in the document]
Comment (jnsbck) [reply jaxley/modules/base.py:540]: I made this change also to acommodate the inhomogenous nsegs we can now have across branches.

File: jaxley/modules/base.py [hunk 5]
```diff
@@ -120,8 +123,137 @@ def __init__(self):
         # `self._init_morph_for_debugging` is run.
         self.debug_states = {}
 
-    def _update_nodes_with_xyz(self):
```
[the remark this answers is not in the document]
Comment (jnsbck) [reply jaxley/modules/base.py:123]: It does in L307

**Comment by jnsbck** [comment]
Awesome! Thanks for the review. Changed a bunch of stuff and left some things open for further discussions.

**Review by michaeldeistler — APPROVED** [review]

**Pull request merged**